File: backend/app/models/account.rb
# == Schema Information
#
# Table name: accounts
#
#  id             :bigint           not null, primary key
#  account_number :string(50)
#  account_type   :string           default("savings"), not null
#  balance        :decimal(14, 2)   default(0.0), not null
#  closed_amount  :decimal(14, 2)
#  closed_date    :date
#  nickname       :string(100)      not null
#  open_date      :date             not null
#  created_at     :datetime         not null
#  updated_at     :datetime         not null
#  bank_id        :bigint           not null
#  user_id        :bigint           not null
#
# Indexes
#
#  index_accounts_on_bank_id  (bank_id)
#  index_accounts_on_user_id  (user_id)
#
# Foreign Keys
#
#  fk_rails_...  (bank_id => banks.id) ON DELETE => restrict
#  fk_rails_...  (user_id => users.id) ON DELETE => cascade
#
class Account < ApplicationRecord
  class Error < StandardError; end

  audited only: [ :balance ]

  has_many :imports, class_name: "ImportBatch", as: :linked_account
  has_many :transactions, as: :linked_account

  belongs_to :user
  belongs_to :bank

  enum :account_type, {
    savings: "savings",
    current: "current",
    salary:  "salary",
    nre:     "nre",
    nro:     "nro"
  }, validate: true

  scope :open,   -> { where(closed_date: nil) }
  scope :closed, -> { where.not(closed_date: nil) }

  validates :nickname,  presence: true
  # `open_date` is the immutable cutoff: anything before it is folded into
  # the user-supplied opening transaction (a regular credit dated on that
  # day). The DB column is NOT NULL so this presence validation is belt
  # + braces.
  validates :open_date, presence: true

  def closed?
    closed_date.present?
  end

  # Direct credit/debit (not driven by a Transaction). Callers MUST pass
  # `source:` so the audit row records what caused the change — e.g.
  # `account.credit!(amount, source: "close:term_account_#{ta.id}")`.
  # Without a source the audit log shows an anonymous "Balance update",
  # which is what we used to do — and which made past mysteries hard to
  # trace back to their cause.
  def debit!(amount, source: nil)
    raise Error, "Account '#{nickname}' is closed" if closed?
    raise Error, "Insufficient balance in '#{nickname}' (available: #{balance}, required: #{amount})" if balance < amount
    self.audit_comment = source if source
    update!(balance: balance - amount)
  end

  def credit!(amount, source: nil)
    raise Error, "Account '#{nickname}' is closed" if closed?
    self.audit_comment = source if source
    update!(balance: balance + amount)
  end
end

File: backend_python/app/models/account.py
import enum
from datetime import date, datetime

from sqlalchemy import Date, DateTime, Enum, ForeignKey, Numeric, String, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base
from app.models._audit_registry import auditable


class AccountType(str, enum.Enum):
    savings = "savings"
    current = "current"
    salary = "salary"
    nre = "nre"
    nro = "nro"


@auditable("balance")
class Account(Base):
    """A user's account at a specific bank — the many-to-many junction."""

    __tablename__ = "accounts"

    id: Mapped[int] = mapped_column(primary_key=True)
    user_id: Mapped[int] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), index=True, nullable=False
    )
    bank_id: Mapped[int] = mapped_column(
        ForeignKey("banks.id", ondelete="RESTRICT"), nullable=False
    )
    nickname: Mapped[str] = mapped_column(String(100), nullable=False)
    account_number: Mapped[str | None] = mapped_column(String(50), nullable=True)
    account_type: Mapped[AccountType] = mapped_column(
        Enum(AccountType, name="account_type"), nullable=False, default=AccountType.savings
    )
    balance: Mapped[float] = mapped_column(Numeric(14, 2), nullable=False, default=0, server_default="0")
    open_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    closed_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    closed_amount: Mapped[float | None] = mapped_column(Numeric(14, 2), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

    def debit(self, amount: float) -> None:
        if self.closed_date:
            raise ValueError(f"Account '{self.nickname}' is closed")
        if float(self.balance) < amount:
            raise ValueError(
                f"Insufficient balance in '{self.nickname}' (available: {self.balance}, required: {amount})"
            )
        self.balance = float(self.balance) - amount

    def credit(self, amount: float) -> None:
        if self.closed_date:
            raise ValueError(f"Account '{self.nickname}' is closed")
        self.balance = float(self.balance) + amount

    bank = relationship("Bank", back_populates="accounts")
    user = relationship("User", back_populates="accounts")
    transactions = relationship(
        "Transaction",
        primaryjoin="and_(Account.id == foreign(Transaction.linked_account_id), Transaction.linked_account_type == 'account')",
        viewonly=True,
    )
    audit_logs = relationship(
        "AuditLog",
        primaryjoin="and_(Account.id == foreign(AuditLog.record_id), AuditLog.table_name == 'accounts')",
        viewonly=True,
    )
